rust/arrow/src/datatypes.rs

```diff
@@ -949,28 +984,59 @@ impl Schema {
     pub fn to_json(&self) -> Value {
         json!({
             "fields": self.fields.iter().map(|field| field.to_json()).collect::<Vec<Value>>(),
+            "metadata": serde_json::to_value(&self.metadata).unwrap()
         })
     }
 
     /// Parse a `Schema` definition from a JSON representation
     pub fn from(json: &Value) -> Result<Self> {
         match *json {
             Value::Object(ref schema) => {
-                if let Some(Value::Array(fields)) = schema.get("fields") {
-                    let fields: Result<Vec<Field>> =
-                        fields.iter().map(|f| Field::from(f)).collect();
-                    Ok(Schema::new(fields?))
+                let fields = if let Some(Value::Array(fields)) = schema.get("fields") {
+                    fields
+                        .iter()
+                        .map(|f| Field::from(f))
+                        .collect::<Result<_>>()?
                 } else {
                     return Err(ArrowError::ParseError(
                         "Schema fields should be an array".to_string(),
                     ));
-                }
+                };
+
+                let metadata = if let Some(value) = schema.get("metadata") {
+                    Self::_from_metadata(value)?
+                } else {
+                    HashMap::default()
+                };
+
+                Ok(Self { fields, metadata })
             }
             _ => Err(ArrowError::ParseError(
                 "Invalid json value type for schema".to_string(),
             )),
         }
     }
+
+    /// Parse a `metadata` definition from a JSON representation
+    fn _from_metadata(json: &Value) -> Result<HashMap<String, String>> {
```

Comment:
We don't prefix functions with `_` if they're private. I'm assuming this is why you did this?

```suggestion
    fn from_metadata(json: &Value) -> Result<HashMap<String, String>> {
```

Reply:
Done